### PR TITLE
fix #274540: save user choice of instrument genre

### PR DESCRIPTION
**Pull request by jthistle**
This addresses a feature request at: https://musescore.org/en/node/274540

The request was to keep the user's selection of instrument genre so that they don't have to change it every time they add/change an instrument. This PR adds this functionality by saving the user's last choice of instrument in QSettings, which actually means that the last selected genre will be saved even after the application has closed.

### CODE REVIEW

File: mscore/instrwidget.cpp [hunk 1]
```diff
@@ -362,6 +362,8 @@ QString InstrumentTemplateListItem::text(int col) const
 InstrumentsWidget::InstrumentsWidget(QWidget* parent)
    : QWidget(parent)
       {
+      storeChangeInGenre = false;
```
Comment (Marr11317) [inline mscore/instrwidget.cpp]: Why is this needed?
Couldn't it be set directly to true?

File: mscore/instrwidget.cpp [hunk 2]
```diff
@@ -385,6 +387,16 @@ InstrumentsWidget::InstrumentsWidget(QWidget* parent)
       linkedButton->setEnabled(false);
 
       connect(instrumentList, SIGNAL(clicked(const QModelIndex &)), SLOT(expandOrCollapse(const QModelIndex &)));
+      
+      // get last user selected instrument genre and set filter to it
+      QSettings settings;
+      if ( ! settings.value("selectedInstrumentGenre").isNull() ){
+            int selectedGenre = settings.value("selectedInstrumentGenre").value<int>();
+            instrumentGenreFilter->setCurrentIndex(selectedGenre);
+            }
```
Comment (Marr11317) [inline mscore/instrwidget.cpp]: these last lines are not needed.
just use 
```cpp
instrumentGenreFilter->setCurrentIndex(settings.value("selectedInstrumentGenre", /*default value, probably 0*/ 0).toInt());
```
same for in SelectInstrument.cpp

File: mscore/instrwidget.cpp [hunk 3]
```diff
@@ -897,6 +911,12 @@ void InstrumentsWidget::on_search_textChanged(const QString &searchPhrase)
 
 void InstrumentsWidget::on_instrumentGenreFilter_currentIndexChanged(int index)
       {
+      QSettings settings;
+      if (storeChangeInGenre){
+            settings.setValue("selectedInstrumentGenre", index);
+            settings.sync();
+            }
```
Comment (Marr11317) [inline mscore/instrwidget.cpp]: couldn't these settings be saved in the destructor?
They would only be saved once, that way.

File: mscore/instrwidget.cpp (hunk 1 above)
Comment (jthistle) [reply mscore/instrwidget.cpp]: This variable is needed because when the InstrumentsWidget is being created, the selected instrument genre is changed by some other scripts a couple of times. I want to be able to save changes to the genre, but only ones made by the user. So, when loading is complete and this variable has been set to true, only then will any genre changes made be stored, because they are user-made changes, not automatic ones.

File: mscore/instrwidget.cpp (hunk 3 above)
Comment (jthistle) [reply mscore/instrwidget.cpp]: That's a good point actually. I can't test it right now, but I seem to remember that if you access the value of a setting after it's been changed but not synced, it will have the last synced value. The variable would be unneeded, I guess because:

1. The widget is initialized 
2. The genre settings value is changed a couple of times automatically but not synced
3. The genre filter is set to the genre settings value (but the last saved one, which would be user chosen)
4. This change in genre filter is saved to the genre settings variable (meaning that any automatic genre changes are basically ignored) 
5. User possibly changes genre
6. Widget closes, last chosen genre is saved

Sounds good to me. Basically, you're right :)

File: mscore/instrwidget.cpp (hunk 1 above)
Comment (jthistle) [reply mscore/instrwidget.cpp]: The variable isn't needed though if you see my reply below I guess

**Comment by jthistle** [comment]
Thanks for your review @Marr11317. Unfortunately I'll be unable to implement any changes for the next 3 weeks, but your feedback is very useful.

**Comment by jthistle** [comment]
Feedback addressed, a final review would be appreciated @Marr11317 (or anyone else)

File: mscore/selinstrument.h [hunk 4]
```diff
@@ -50,6 +50,7 @@ class SelectInstrument : public QDialog, private Ui::SelectInstrument {
    public:
       SelectInstrument(const Instrument*, QWidget* parent = 0);
       const InstrumentTemplate* instrTemplate() const;
+      ~SelectInstrument();
       };
```
Comment (Marr11317) [inline mscore/selinstrument.h]: same as [upper](https://github.com/musescore/MuseScore/pull/3830#pullrequestreview-146002262)

File: mscore/instrwidget.cpp [hunk 5]
```diff
@@ -1052,4 +1062,15 @@ QTreeWidget* InstrumentsWidget::getPartiturList()
...
+
+//---------------------------------------------------------
+//   destructor
+//---------------------------------------------------------
+
+InstrumentsWidget::~InstrumentsWidget()
+      {
+      QSettings settings;
+      settings.sync();
+      }
```
[the remark this answers is not in the document]
Comment (jthistle) [reply mscore/instrwidget.cpp]: Thanks for the review. You're right, the settings value could be written in the destructor. I'm a bit new to C++ so I don't know the ins and outs of it very well yet, so your feedback really helps :)

**Comment by Marr11317** [comment]
Here's a few things you could also do around:
 - remove the `InstrumentsWidget::writeSettings()` function, since it seems like it has no definition, and is never called.
 - replace the few Qt's foreach by c++ 11's for in instrwidget.cpp.

**Comment by jthistle** [comment]
@Marr11317 

> Actually, I think the index should not be saved as an int, but as a QString

I don't think it should because, firstly, I don't think `setCurrentText()` is a method of the `instrumentGenreFilter`, so it would have to be written. Secondly, because you could have the reverse of the problem you're talking about: you're worried that instrument genre indices may change, but what about the instrument genre names? Surely those could change as well, and if they do then there would be a similar problem. Because of this, I'm going to stick with saving the index.

**Comment by jthistle** [comment]
Any final review needed @Marr11317 ?

**Comment by Marr11317** [comment]
> you're worried that instrument genre indices may change, but what about the instrument genre names? Surely those could change as well, and if they do then there would be a similar problem

But then, if the string changes, and so the saved string is not found, the index will be set to 0, which means general instruments. But if new instruments lists are added, and you try to put the index to let's say 3, It could be something completely random such as "Indian folk instruments of the nineteenth century"

**Comment by jthistle** [comment]
@Marr11317 

The whole point of this feature request is so that the genre is the same. It doesn't make sense to have it different in different dialogues. The genre, when changed in the 'add instrument' windows, should also change in the 'change instrument' window (for changing the instrument of a stave). The reasoning being that when working on a project with a particular genre of instruments you don't want to have to keep changing the genre, neither when you're adding an instrument nor when you're changing one. So, the value *must* be saved in the same place, something that can't be done across objects using the `objectName()` as the group name.

**Comment by Marr11317** [comment]
looks good to me.

**Comment by jthistle** [comment]
@Marr11317 after some investigation, I've found out that things aren't as simple as they might be :/

The instruments widget doesn't ever destruct during runtime - it constructs once, when first opened, then destructs once when the program closes. The select instrument window, on the other hand, *does* construct every time when opened, and destructs every time it's closed. I'm going to have to go back to writing and syncing in the `on_instrumentGenreFilter_currentIndexChanged` method, as inelegant as it may be. This may require me having to use a variable again, but hey, what can you do sometimes :man_shrugging:

**Comment by jthistle** [comment]
@Marr11317 

Right. This should be the final commit. Things have had to be changed a bit - as I said before, the Instrument Dialog wasn't being destroyed on close. Now, it is destroyed on close, but it's not simple, because sometimes the Instrument Widget (a child of the Dialog) needs to persist for a while after closure to allow instrument changes to be made. So the, Dialog needs to be closed at different points in the code depending on if a change has been made or not (and so it needs to happen in the `editInstrList` method). And, because the Dialog is a child of the main window, cleanup needs to happen in the main MuseScore class, which is why I've added a method to it in the musescore.cpp file. Phew. Saving does now happen in destructors though, which is what you wanted I think :)

File: mscore/musescore.cpp [hunk 6]
```diff
@@ -5956,6 +5956,19 @@ QMenu* MuseScore::createPopupMenu()
...
 
+//---------------------------------------------------------
+//   closeInstrumentsDialog
+//---------------------------------------------------------
+
+void MuseScore::closeInstrumentsDialog()
+      {
+      if (instrList){
+            // ugly, but instrList->done(1) doesn't work, so instrList must be directly destroyed
+            instrList->~InstrumentsDialog();
```
Comment (Marr11317) [inline mscore/musescore.cpp]: what about `delete instrList;`

Comment (jthistle) [reply mscore/musescore.cpp]: ah, yes that would be better 🤦‍♂️

**Comment by jthistle** [comment]
There, all done @Marr11317 . Ready for merge now @anatoly-os ?

**Comment by Jojo-Schmitz** [comment]
Rebase needed, conflict should not be too difficult to fix

**Comment by jthistle** [comment]
Accidently force-pushed branch without committing :man_facepalming: 
Reopening.

**Comment by jthistle** [comment]
This PR is outdated. The code is also badly written. I will rewrite it, and open a new PR, and as such I am closing this one.
